### ISSUE TITLE
error: more than one emulator

### ISSUE THREAD

**Issue by Heavedistant** [Egoistically/ALAuto#37]
Hello, trying to use this amazing script again and running into error: more than one emulator again.
I also tried the -l legacy parameter to see if SED would work. Any ideas what I might have wrong?

Without -l:
```
[2019-12-15 02:16:34] Initializing config module
[2019-12-15 02:16:34] Validating config
[2019-12-15 02:16:34] Starting ALAuto!
* daemon not running; starting now at tcp:5037
* daemon started successfully
[2019-12-15 02:16:40] Successfully connected to the service.
[2019-12-15 02:16:59] Found specified map.
[2019-12-15 02:17:11] Started map clear.
error: more than one emulator
Traceback (most recent call last):
  File "ALAuto.py", line 197, in <module>
    script.run_sortie_cycle()
  File "ALAuto.py", line 72, in run_sortie_cycle
    self.run_combat_cycle()
  File "ALAuto.py", line 80, in run_combat_cycle
    result = self.modules['combat'].combat_logic_wrapper()
  File "C:\Users\Games\Desktop\ALAuto-master\modules\combat.py", line 86, in combat_logic_wrapper
    if not self.clear_map():
  File "C:\Users\Games\Desktop\ALAuto-master\modules\combat.py", line 357, in clear_map
    target_info = self.get_closest_target(self.blacklist)
  File "C:\Users\Games\Desktop\ALAuto-master\modules\combat.py", line 510, in get_closest_target
    fleet_location = self.get_fleet_location()
  File "C:\Users\Games\Desktop\ALAuto-master\modules\combat.py", line 470, in get_fleet_location
    Utils.update_screen()
  File "C:\Users\Games\Desktop\ALAuto-master\util\utils.py", line 84, in update_screen
    screen = cv2.imdecode(numpy.fromstring(Adb.exec_out('screencap -p'), dtype=numpy.uint8), 0)
cv2.error: OpenCV(3.4.4) C:\projects\opencv-python\opencv\modules\imgcodecs\src\loadsave.cpp:750: error: (-215:Assertion failed) !buf.empty() && buf.isContinuous() in function 'cv::imdecode_'
```

With -l:
```
[2019-12-15 02:17:35] Initializing config module
[2019-12-15 02:17:35] Validating config
[2019-12-15 02:17:35] Starting ALAuto!
[2019-12-15 02:17:35] Enabled sed usage.
* daemon not running; starting now at tcp:5037
* daemon started successfully
[2019-12-15 02:17:41] Successfully connected to the service.
error: more than one emulator
Traceback (most recent call last):
  File "ALAuto.py", line 184, in <module>
    Utils.update_screen()
  File "C:\Users\Games\Desktop\ALAuto-master\util\utils.py", line 82, in update_screen
    screen = cv2.imdecode(numpy.fromstring(Adb.exec_out(r"screencap -p | sed s/\r\n/\n/"),dtype=numpy.uint8),0)
cv2.error: OpenCV(3.4.4) C:\projects\opencv-python\opencv\modules\imgcodecs\src\loadsave.cpp:750: error: (-215:Assertion failed) !buf.empty() && buf.isContinuous() in function 'cv::imdecode_'
```

**Comment by hjk22** [comment]
It seems the adb call fails because the daemon detects more than one emulator and doesn't know which one it needs to interact with, it shouldn't be related to the usage of SED since the cv2.error seems to report that it found an empty buffer.
When you get this error, open the cmd and type this command:
`adb devices -l`
which lists the devices currently attached to the adb daemon, and post its output here.
Also if you could give us more information about your setup (like adb version, emulator used), that would be helpful to us in order to figure out the cause of this issue.

**Comment by Heavedistant** [comment]
Thanks for the response @hjk22 . I'm running LDPlayer as my emulator which worked great the last time I was running this. `adb devices -l` shows the 1 emulator as expected.
```
C:\Users\Games\Desktop\ALAuto-master>adb devices -l
List of devices attached
emulator-5556          device product:oppo a33m model:oppo_a33m device:x86 transport_id:2
```

ADB Version
```
Android Debug Bridge version 1.0.41
Version 29.0.5-5949299
```
OS `Windows 10`
Python `3.6.5`

```
C:\Users\Games\Desktop\ALAuto-master>py ALAuto.py
[2019-12-16 00:14:04] Initializing config module
[2019-12-16 00:14:04] Validating config
[2019-12-16 00:14:04] Starting ALAuto!
* daemon not running; starting now at tcp:5037
* daemon started successfully
[2019-12-16 00:14:10] Successfully connected to the service.
[2019-12-16 00:14:43] Current oil: 4661
[2019-12-16 00:14:43] Combat done: 0 / attempted: 0
[2019-12-16 00:14:43] ALAuto has been running for 0 hours 0 minutes 39 seconds (started on 2019-12-16 00:14:04)
error: more than one emulator
Traceback (most recent call last):
  File "ALAuto.py", line 197, in <module>
    script.run_sortie_cycle()
  File "ALAuto.py", line 72, in run_sortie_cycle
    self.run_combat_cycle()
  File "ALAuto.py", line 80, in run_combat_cycle
    result = self.modules['combat'].combat_logic_wrapper()
  File "C:\Users\Games\Desktop\ALAuto-master\modules\combat.py", line 73, in combat_logic_wrapper
    Utils.wait_update_screen(1)
  File "C:\Users\Games\Desktop\ALAuto-master\util\utils.py", line 97, in wait_update_screen
    cls.update_screen()
  File "C:\Users\Games\Desktop\ALAuto-master\util\utils.py", line 84, in update_screen
    screen = cv2.imdecode(numpy.fromstring(Adb.exec_out('screencap -p'), dtype=numpy.uint8), 0)
cv2.error: OpenCV(3.4.4) C:\projects\opencv-python\opencv\modules\imgcodecs\src\loadsave.cpp:750: error: (-215:Assertion failed) !buf.empty() && buf.isContinuous() in function 'cv::imdecode_'
```

**Comment by hjk22** [comment]
That's strange. The output of your `adb devices -l` is bugging me because the device attached has "2" as transport_id, making me think that there was another one attached which got disconnected.

I tried looking through the adb's issue tracker kept by Google, but I didn't find one that fits our scenario.

Since you said that you got the bot working with an older adb version, you could try to revert to an older version, if you want to maintain your current setup. I'm thinking that LDPlayer might have some problems interacting with newer versions of adb (just making an assumption based on the little info I could gather, I'm not familiar with this emulator). Check if you are using the latest version of your emulator available, maybe the devs were aware of similar issues and fixed them. 
If this doesn't work, you might want to consider changing emulator.

Another workaround that I could come up with is for you to change the code  of `adb.py`: substitute every Adb.device and self.device inside of lists with `'-s', 'SERIALNUMBER'` (from your `adb devices` output, the serial number of your device is `emulator-5556`).
This way, even if your daemon detects more than one device, it shouldn't fail because there won't be any ambiguity, since you are specifying which device it should interact with.
If you get `error: device not found`, comment the lines which starts and kills the server from `adb.py` and then, before launching the bot, start the adb server manually yourself and run the command `adb devices` (typing this command should actually start the server, if it isn't already running) to get the current emulator's serial number and put that into the code.
This doesn't solve the issue, but it should let you run the script.

Sorry for not being more helpful, but these are what I can think up right now.

**Comment by Heavedistant** [comment]
Sorry for the delay in following up. Thanks again for looking into the issue again. I can see ALAuto interact with LDPlayer and read the oil amount, click on some buttons and then hit a Python error. So it seems like it could be a configuration issue between adb and other libraries or something. I'm going to try the serial number method you suggested, that sounds very interesting.
I appreciate you taking the time to troubleshoot this for me. I will try some things and follow up with how it goes!

**Comment by hjk22** [comment]
You're welcome! If you have any updates on this matter let me know.

**Comment by Heavedistant** [comment]
Following up, I tried the -s serialnumber edit and it does prevent an error but seems to just hang afterward with no further output. The funny thing is that this was all working flawlessly with the -l SED mode that was added for legacy support. I think I'm going to try to back ADB versions, etc and see if I can get my setup more stable. Thank you again for the suggestions and continued work maintaining this repo.
